### PR TITLE
Inconsistencies spec

### LINKED ISSUE #3583: Inconsistencies spec
Need to pull together a spec covering what inconsistencies currently exist for our renderers, along with a plan for addressing them and preventing them in the future.

### PR DESCRIPTION
**Pull request by paulcam206**
## Related Issue
Fixes #3583

## Description
Draft spec for dealing with inconsistencies. Probably not quite finished yet, but complete enough to begin discussions.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/3633)

### CODE REVIEW

File: specs/features/Inconsistencies.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+* Supporting libraries and platform affordances available to us can ease or speed up implementation, but can also be the source of inconsistencies (e.g. differences in how a dev constructs .NET's `DateTime`, UWP's `DateTime`, C++'s `std::chrono::time_point`, and Javascript's `Date` -- instantiating each object with, e.g a string can vary in permissivity)
+  * Indeed, we've had a few issues arise from differences between a hand-rolled solution and a pre-rolled solution (e.g. date/time handling (#1915) and support for markdown (#1984))
+
+### Rendering
+
+Rendering presents another avenue by which inconsistencies can sneak into Adaptive Cards. Some of these problems can be pretty tricky to solve.
+
+* Multiple implementations means that "gotcha" scenarios might not always be covered or may be covered inconsistently (e.g. what happens when you combine `maxLines`, markdown lists, and `wrap: true`?)
+* Platform-provided controls can vary widely in their behaviors and support. Even if divergent behavior is caught during implementation, it may be difficult or tricky to solve completely
+* Some
```
Comment (RebeccaAnne) [inline specs/features/Inconsistencies.md]: Incomplete :)

Comment (paulcam206) [reply specs/features/Inconsistencies.md]: hahaha... indeed 👍

File: specs/features/Inconsistencies.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+* [Javascript](https://docs.microsoft.com/en-us/adaptive-cards/sdk/rendering-cards/javascript/getting-started) (renders using HTML/JS/CSS)
+* [ReactNative](https://docs.microsoft.com/en-us/adaptive-cards/sdk/rendering-cards/react-native/getting-started) (community renderer)
+* Flutter (community renderer)
+
+This list of platforms represents at least 7 distinct implementations of Adaptive Cards at the rendering level. At each decision point along the way during implementation, the Adaptive Cards developer (i.e. someone writing code to parse or render Adaptive Cards) risks implementing behavior that diverges from intended behavior. One of the biggest selling points for Adaptive Cards is our ability to render across our supported platforms in a way that at once feels native to the supported platform while also having a look and feel that's representative of the card author's intent. Indeed, portability is one of our [core goals](https://docs.microsoft.com/en-us/adaptive-cards/#goals). Each time an inconsistency sneaks into the platform, it risks introducing costs to our partners. These costs have real consequences for the adoption of Adaptive Cards in the market:
```
Comment (shalinijoshi19) [inline specs/features/Inconsistencies.md]: It'd be nice to list out specifically teh goals and non-goals of the work you are setting up to do in this spec to scope it a bit. For eg we are not attempting to solve the inconsistencies that occur across iOS and Android renderers being customized by a given partner. In addition the specific differences between Android and iOS in particular (so the intra-mobile) inconsistencies need not be tackled here; They are being driven by Joseph (@nesalang) and perhaps can be referenced in an Appendix or References section here but clarify the scope of this document doesnt address that category. ANd so on..

File: specs/features/Inconsistencies.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+2
```
Comment (shalinijoshi19) [inline specs/features/Inconsistencies.md]: Take a look at a couple of specs that can help structure the narrative here a bit maybe -  the first is more high level if thats where you want to go (state of affairs today fr eg) and the 2nd is a bit more "closer to the ground" :) let me know what you think/thoughts-- ([https://microsoft.sharepoint.com/teams/WDGProjectRome/Shared%20Documents/Project%20Rome%20Reliability%20Review.docx?web=1](https://microsoft.sharepoint.com/teams/WDGProjectRome/Shared%20Documents/Project%20Rome%20Reliability%20Review.docx?web=1)) is one; there's another one around this [https://microsoft.sharepoint.com/teams/specstore/Developer%20Platform%20Team%20DEP/Rome%20Reliability%20Spec.docx?web=1](https://microsoft.sharepoint.com/teams/specstore/Developer%20Platform%20Team%20DEP/Rome%20Reliability%20Spec.docx?web=1)

File: specs/features/Inconsistencies.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  * Make sure we only break our customers once
+* Inconsistency-focused bug bash
+
+## Specific bugs
+
+[List of inconsistency bugs](https://github.com/microsoft/AdaptiveCards/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3AInconsistency+):
+
+* [3605](https://github.com/microsoft/AdaptiveCards/issues/3605) - [Android][Rendering Discrepancy between Android and web] [For Choice sets with compact style, when no option is explicitly pre-selected, first option gets pre-selected]
+* [3573](https://github.com/microsoft/AdaptiveCards/issues/3573) - [.NET C#][Input.ChoiceSet] [adaptive card dropdown title with double quotes rendering issue]
+* [3572](https://github.com/microsoft/AdaptiveCards/issues/3572) - [Web][Avatars] Person avatars are not always circular
```
Comment (shalinijoshi19) [inline specs/features/Inconsistencies.md]: It would be nice to bucketize these issues and include them in the respective inconsistency bucket they belong in above that you are identifying as categories

File: specs/features/Inconsistencies.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+### Rendering
+
+Rendering presents another avenue by which inconsistencies can sneak into Adaptive Cards. Some of these problems can be pretty tricky to solve.
+
+* Multiple implementations means that "gotcha" scenarios might not always be covered or may be covered inconsistently (e.g. what happens when you combine `maxLines`, markdown lists, and `wrap: true`?)
+* Platform-provided controls can vary widely in their behaviors and support. Even if divergent behavior is caught during implementation, it may be difficult or tricky to solve completely
+* Some 
+
+## Proposals
+
```
Comment (shalinijoshi19) [inline specs/features/Inconsistencies.md]: Are we saying that we are going with all of these? some of these? all or nothing? be nice if we could present these in order of the end goal/cost/prioritization etc Which ones are low hanging fruit and must haves and to what end? What category do they help address?

File: specs/features/Inconsistencies.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+* [ReactNative](https://docs.microsoft.com/en-us/adaptive-cards/sdk/rendering-cards/react-native/getting-started) (community renderer)
+* Flutter (community renderer)
+
+This list of platforms represents at least 7 distinct implementations of Adaptive Cards at the rendering level. At each decision point along the way during implementation, the Adaptive Cards developer (i.e. someone writing code to parse or render Adaptive Cards) risks implementing behavior that diverges from intended behavior. One of the biggest selling points for Adaptive Cards is our ability to render across our supported platforms in a way that at once feels native to the supported platform while also having a look and feel that's representative of the card author's intent. Indeed, portability is one of our [core goals](https://docs.microsoft.com/en-us/adaptive-cards/#goals). Each time an inconsistency sneaks into the platform, it risks introducing costs to our partners. These costs have real consequences for the adoption of Adaptive Cards in the market:
+2
+* Existing customers are less likely to adopt new versions of Adaptive Cards, even if they want the newer features
```
Comment (shalinijoshi19) [inline specs/features/Inconsistencies.md]: is there evidence to support this? We should include existing pain points/issues that have caused us to take this on to help lay out the context for the reader/reviewer here a bit. What's teh pain points today that we know of?

File: specs/features/Inconsistencies.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+* Potential new customers might be scared to adopt Adaptive Cards altogether if there's a consistency or stability problem (perceived or real), since cross-platform capabilities are central to our value proposition
+* When existing customers *do* adopt new versions, they may end up having to spend extra time testing new versions and then deal with new bugs when *their* customers complain
```
Comment (shalinijoshi19) [inline specs/features/Inconsistencies.md]: So again, we are beyond the point of "may"/hypotheticals at this point which is why we are now tackling this and prioritizing this work head on. We should delve straight into that - a discussion of existing pain points vs the hypothetical

File: specs/features/Inconsistencies.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+### In order to maintain our pledge of cross-platform capability, and in order to support all of our partners and customers, we **must** take a strong stance against the introduction of inconsistencies in Adaptive Cards
```
Comment (shalinijoshi19) [inline specs/features/Inconsistencies.md]: > pledge of cross-platform capabilit [](start = 28, length = 35)

Elaborate a bit more to be be specific and include pointers to actual public docs or some such here. Also "pledge" seems a bit dramatic :) - Can stick to "goals" or vision as outlined [add link]; In addition customers, esp those spanning product offerings spanning multiple platforms have all reported numerous bugs and pain points esp as they upgrade or expand adaptive cards usage across these platforms (link to the github issues here reported by teams/cortana/teams desktop/bot framework)

File: specs/features/Inconsistencies.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+Rendering presents another avenue by which inconsistencies can sneak into Adaptive Cards. Some of these problems can be pretty tricky to solve.
+
+* Multiple implementations means that "gotcha" scenarios might not always be covered or may be covered inconsistently (e.g. what happens when you combine `maxLines`, markdown lists, and `wrap: true`?)
+* Platform-provided controls can vary widely in their behaviors and support. Even if divergent behavior is caught during implementation, it may be difficult or tricky to solve completely
+* Some 
+
+## Proposals
+
+* Attempt to reduce code duplication:
```
Comment (nesalang) [inline specs/features/Inconsistencies.md]: We could also give more options to developers, so they can customize their experiences. for example, we can allow custom markdown library to all non-js libraries as js library does already. In that way, when they file an osistency issue, we can cater them to use custom library. markdown is well standardized, it should be easy for them to find one. I think this will work only if js libary has access to the shared markdown library since the shared markdown libary is the least common denominator.

File: specs/features/Inconsistencies.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  * Experiment with the shared model as Javascript (e.g. via [emscripten](https://emscripten.org/))
+  * Move .NET from using Newtonsoft.JSON to using the shared model via P/Invoke (Note that this will have impacts on partners. It's likely that this move will only make sense if Javascript is also using the shared model)
+* Changes processes to prevent new inconsistencies from ever happening:
+  * New bugs automatically filed with `inconsistency` tag, which is only removed once it's determined that the issue is *not* a consistency issue
+  * We can't ship if an `inconsistency`-tagged bug hasn't been resolved for this release
+  * Features that need implementation across all platforms are merged to a single branch and PR'd/tested for inconsistencies before being allowed to merge upstream
```
Comment (nesalang) [inline specs/features/Inconsistencies.md]: i like this idea. can we also pick a desinated designer (or an architect) per the feature? and require the chosen designer to present design to the design meeting with minimum of a prototype and brief overview of the design. After the review approval, the designer implement the feature on at least one platform and complete document on how the feature works. Then other devs pick up the work on the remaining platforms. Designers can be anyone. 
This approach will save dev times by not having to design features each time working on a different platform. Although platforms are different, our feature is common across the all platforms. Feature implementation design doesn't need to be re-desined, and this also allow the plaform specialzation to benefit us. If we are tight on schedule, we can assign feature developement on iOS to the dev with iOS specialization. If we have leeway on schedule, give to dev who wants to learn about a platform. 

Sometimes a plaform can be too different that the design can not be applied directly, for exaple, if negative margin is used in .NET, but iOS can't use negative margin. Even such case, the implementor doesn't need to start from scratch, and the design can serve as scafolding for the dev's own implementation. This also gives opportunity to document. Documenting the difference is much easier than writing a whole new document.

it's aslo benetfitial in maintanence perspective. Since we have several implementors of a feature, and if the feature works the same way, we could assign any devs to work on any platforms (yes, this is a little bit of exaggeration). There will be differences, but the consistent feature implementation will serve as bread crumbs to navigate through differences in platforms, and it will save time in debugging.

**Comment by VikrantSingh01** [comment]
Please add this one too for new line behavipur for empty text block Across web and mobile.
{
          "type": "TextBlock",
          "text": ""
},
Link : https://github.com/microsoft/AdaptiveCards/issues/3702

**Comment by paulcam206** [comment]
#3702 is effectively a duplicate of #641, which is in the list :)

---
In reply to: [571427833](https://github.com/microsoft/AdaptiveCards/pull/3633#issuecomment-571427833) [](ancestors = 571427833)

File: specs/DesignDiscussions/Inconsistencies.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+* Platform-provided controls can vary widely in their behaviors and support. Even if divergent behavior is caught during implementation, it may be difficult or tricky to solve completely
+
+## Proposals
+
+* (P0) Change processes to prevent new inconsistencies from ever happening:
+  * Motivation: Inconsistencies are caught before they ship
+  * Tasks: 
+    * New bugs automatically filed with `inconsistency` tag, which is only removed once it's determined that the issue is *not* a consistency issue
+    * We can't ship if an `inconsistency`-tagged bug hasn't been resolved for this release
+    * Features that need implementation across all platforms are merged to a single branch and PR'd/tested for inconsistencies before being allowed to merge upstream
```
Comment (RebeccaAnne) [inline specs/DesignDiscussions/Inconsistencies.md]: I like this. We can do bug bashes on these branches too.

File: specs/DesignDiscussions/Inconsistencies.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+* (P1) Provide a new tool that allows easy testing of card payloads across all renderers:
+  * Motivation: Allows for the easy directed testing of feature consistency
+  * Tasks: 
+    * If shipped for public use, this could be a nice addition to the designer to allow card authors to preview their cards in a variety of renderers
+    * Could be used in tests for per-platform visual verification
+    * Per-platform accessibility trees could also be generated by this tool to allow for cross-plat verification
+* (P1) Make hard decisions around defaults, required properties, and parser/renderer behaviors: 
+  * Motivation: When these decisions are fully documented, it's easier for a line-level dev to choose correct behavior, meaning that it's more intuitive as to what the correct behavior should be.
+  * Tasks: 
+    * Document decisions publicly & in code as appropriate
```
Comment (RebeccaAnne) [inline specs/DesignDiscussions/Inconsistencies.md]: > Document decisions publicly & in code as appropriate [](start = 5, length = 53)

Maybe add a note here about updating the schema files in the repro along with the spec and/or code to ensure that decisions make it into the schema documentation for future reference.

File: specs/DesignDiscussions/Inconsistencies.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
+# Inconsistencies in Adaptive Cards platforms
+
+## Motivation
+
+### Goals
+
+* Provide a platform for experiences that allows card authors to write a card once and be confident that the card works everywhere
```
Comment (paulcam206) [inline specs/DesignDiscussions/Inconsistencies.md]: clarify "works"

Comment (paulcam206) [inline specs/DesignDiscussions/Inconsistencies.md]: add API functionality-level coverage

Comment (paulcam206) [reply specs/DesignDiscussions/Inconsistencies.md]: 30k ft view section

File: specs/DesignDiscussions/Inconsistencies.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+* Supporting libraries and platform affordances available to us can ease or speed up implementation, but can also be the source of inconsistencies (e.g. differences in how a dev constructs .NET's `DateTime`, UWP's `DateTime`, C++'s `std::chrono::time_point`, and Javascript's `Date` -- instantiating each object with, for example, a string, can vary in permissivity)
+  * Indeed, we've had a few issues arise from differences between a hand-rolled solution and a pre-rolled solution (e.g. date/time handling (#1915) and support for markdown (#1984))
+
+### Rendering
+
+Rendering presents another avenue by which inconsistencies can sneak into Adaptive Cards. Some of these problems can be pretty tricky to solve.
+
+* Multiple implementations means that "gotcha" scenarios might not always be covered or may be covered inconsistently (e.g. what happens when you combine `maxLines`, markdown lists, and `wrap: true`?)
+* Platform-provided controls can vary widely in their behaviors and support. Even if divergent behavior is caught during implementation, it may be difficult or tricky to solve completely
```
Comment (RebeccaAnne) [inline specs/DesignDiscussions/Inconsistencies.md]: >vary widely [](start = 33, length = 11)

Discussion in the room: Link to Josephs brownbag about ios difficulties

File: specs/DesignDiscussions/Inconsistencies.md (hunk 11 above)
Comment (RebeccaAnne) [inline specs/DesignDiscussions/Inconsistencies.md]: >branch [](start = 84, length = 6)

Possible follow discussion about release branching strategy

File: specs/DesignDiscussions/Inconsistencies.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    * We can't ship if an `inconsistency`-tagged bug hasn't been resolved for this release
+    * Features that need implementation across all platforms are merged to a single branch and PR'd/tested for inconsistencies before being allowed to merge upstream
+      * Each feature is championed by a single dev or PM, who owns the feature from conception to first ship
+      * Champion creates a branch off of `master` (or whichever main branch is appropriate), naming it `feature/featureName` (let's call this the "feature branch")
+      * The champion is responsible for keeping the feature branch up to date with respect to `master` (or main branch)
+      * Workflow:
+        1. People working on the feature create a new branch off of the feature branch (let's call it the "working branch")
+        1. When feature work in a working branch is completed, it is PR'd to the feature branch
+        1. **The first PR to feature branch must include robust test cards that not only exercise the "golden path" for a feature, but every reasonable corner case. This PR must also include schema docs.**
```
Comment (RebeccaAnne) [inline specs/DesignDiscussions/Inconsistencies.md]: Note from Matt in the room: Link to docs on how to generate schema.

File: specs/DesignDiscussions/Inconsistencies.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+
+## Proposals
+
+* (P0) Change processes to prevent new inconsistencies from ever happening:
+  * Motivation: Inconsistencies are caught before they ship
+  * Tasks: 
+    * New bugs automatically filed with `inconsistency` tag, which is only removed once it's determined that the issue is *not* a consistency issue
+    * We can't ship if an `inconsistency`-tagged bug hasn't been resolved for this release
+    * Features that need implementation across all platforms are merged to a single branch and PR'd/tested for inconsistencies before being allowed to merge upstream
+      * Each feature is championed by a single dev or PM, who owns the feature from conception to first ship
```
Comment (RebeccaAnne) [inline specs/DesignDiscussions/Inconsistencies.md]: >single dev or PM [](start = 40, length = 16)

Comment from Shalini in the room: Possibly remove discipline.

Comment (RebeccaAnne) [inline specs/DesignDiscussions/Inconsistencies.md]: >feature [](start = 13, length = 7)

Discussion in the room: More specific discussion of what counts as a "feature" for this process. We don't need feature branches for platform specific features, bugs, etc.

File: specs/DesignDiscussions/Inconsistencies.md (hunk 15 above)
Comment (RebeccaAnne) [inline specs/DesignDiscussions/Inconsistencies.md]: >cards [](start = 69, length = 5)

Discussion in the room: Call out where this card goes (\test) and that we also need a nice looking card for the feature to be shown on the webpage (in elements). Maybe a scenario card too.
